src/librustdoc/doctest.rs

```diff
@@ -833,12 +851,39 @@ fn run_test(
     } else {
         cmd.output()
     };
+
+    // FIXME: Make `test::get_result_from_exit_code` public and use this code instead of this.
+    //
+    // On Zircon (the Fuchsia kernel), an abort from userspace calls the
+    // LLVM implementation of __builtin_trap(), e.g., ud2 on x86, which
+    // raises a kernel exception. If a userspace process does not
+    // otherwise arrange exception handling, the kernel kills the process
+    // with this return code.
+    #[cfg(target_os = "fuchsia")]
+    const ZX_TASK_RETCODE_EXCEPTION_KILL: i32 = -1028;
+    // On Windows we use __fastfail to abort, which is documented to use this
+    // exception code.
+    #[cfg(windows)]
+    const STATUS_FAIL_FAST_EXCEPTION: i32 = 0xC0000409u32 as i32;
+    #[cfg(unix)]
+    const SIGABRT: std::ffi::c_int = 6;
     match result {
         Err(e) => return (duration, Err(TestFailure::ExecutionError(e))),
         Ok(out) => {
-            if langstr.should_panic && out.status.success() {
-                return (duration, Err(TestFailure::UnexpectedRunPass));
-            } else if !langstr.should_panic && !out.status.success() {
+            if langstr.should_panic {
+                match out.status.code() {
+                    Some(test::ERROR_EXIT_CODE) => {}
+                    #[cfg(windows)]
+                    Some(STATUS_FAIL_FAST_EXCEPTION) => {}
+                    #[cfg(unix)]
+                    None if out.status.signal() == Some(SIGABRT) => {}
```

Comment:
I know that you copied this from `libtest` since you told me but I'm a bit unsure about accepting aborts unconditionally even if `-Cpanic=unwind`.

I wondering if it would be more correct to only accept these (SIGABRT, …) if `-Cpanic=abort`. Similarly, I'm wondering if under `-Cpanic=abort` an exit code of 101 would not be deemed acceptable for `should_panic`.

It's 3AM for me, so I can't really judge this right now. I'm just a bit uncomfortable with the idea of immediately shipping these changes given that the scope of this "bug fix" PR has grown in scope. It feels like there are open design questions, I could be wrong tho and I don't want to block this on an FCP.

Reply:
E.g., in the following examples, all tests pass IINM:

```rs
//@ compile-flags: --test -Cpanic=unwind

//! ```should_panic
//! std::process::abort();
//! ```
```

```rs
//@ compile-flags: --test -Cpanic=abort

//! ```should_panic
//! std::process::exit(101);
//! ```
```

Should they? Feels kinda weird to me tbh but maybe I'm the only one.

Reply:
To be transparent, I don't know how `libtest`'s `--test -Cpanic=abort -Zpanic_abort_tests` works under the hood & is supposed to work. It seems to be able to detect aborting `panic!()`s but **also recognize that `std::process::abort()` is not a panic**! No clue how it does that, I have no time to investigate rn.